src/php/Lang/Hasher.php

```diff
@@ -7,67 +7,85 @@
 use RuntimeException;
 
 use function gettype;
-use function is_float;
 use function is_int;
 use function is_object;
+use function is_scalar;
 use function is_string;
 
 /**
  * This Hasher is inspired by the Clojurescript implementation.
+ * These constants are the same hash value as in clojure.
  */
 class Hasher implements HasherInterface
 {
+    private const NULL_HASH_VALUE = 0;
+    private const TRUE_HASH_VALUE = 1231;
+    private const FALSE_HASH_VALUE = 1237;
+    private const POSITIVE_INF_HASH_VALUE = 2146435072;
+    private const NEGATIVE_INF_HASH_VALUE = -1048576;
+    private const DEFAULT_FLOAT_HASH_VALUE = 2146959360;
+
     /**
      * @param mixed $value The value to hash
      *
      * @return int The hash of the given value
      */
     public function hash(mixed $value): int
     {
+        if ($value === null) {
+            return self::NULL_HASH_VALUE;
+        }
+
         if ($value instanceof HashableInterface) {
             return $value->hash();
         }
 
+        if (is_scalar($value)) {
+            return $this->hashScalar($value);
+        }
+
+        if (is_object($value)) {
+            return crc32(spl_object_hash($value));
+        }
+
+        throw new RuntimeException('This type is not hashable: ' . gettype($value));
+    }
+
+    private function hashScalar(float|bool|int|string $value): int
+    {
         if ($value === true) {
-            return 1231; // Same hash value as in clojure
+            return self::TRUE_HASH_VALUE;
         }
 
         if ($value === false) {
-            return 1237; // Same hash value as in clojure
+            return self::FALSE_HASH_VALUE;
         }
 
         if (is_string($value)) {
             return crc32($value);
         }
 
-        if ($value === null) {
-            return 0;
-        }
-
         if (is_int($value)) {
             return $value;
         }
```

Comment:
I know @Chemaclass  your changes are external to this suggestion, but I wonder:

@jenshaase should we contemplate the scenario where an integer number is `INF` or `-INF` as we do with `float` numbers?

Reply:
That's a great observation. I thought about that, but I didn't want to change that logic, I wanted to focus on just refactoring the existing code without changing the existing logic. Although that's a valid point that we could address in a follow up PR. I would say this decision is up to @jenshaase 😄